components/support/rs_client/src/http_client.rs

```diff
@@ -37,53 +95,111 @@ pub(crate) enum RemoteState {
 
 impl Client {
     /// Create a new [Client] with properties matching config.
-    pub fn new(config: ClientConfig) -> Result<Self> {
+    /// During construction, the [Client] will fetch the base attachments
+    /// URL from the server's metadata, assuming attachments are enabled for
+    /// the server.
+    pub fn new(config: RemoteSettingsConfig) -> Result<Self> {
         let server_url = config
             .server_url
             .unwrap_or_else(|| String::from("https://firefox.settings.services.mozilla.com"));
         let bucket_name = config.bucket_name.unwrap_or_else(|| String::from("main"));
+        let base_url = Url::parse(&server_url)?;
+        let lookup = Self::lookup_attachments_base_url(base_url.clone());
```

Comment:
I definitely hear you on sidestepping the complexity of mutexes and optionals to do this the first time an attachment is downloaded! 😅

There are a few reasons why I'm nervous about looking up the base URL at client creation time, though:

* `Client::new` now hides a side effect. It's definitely a personal preference, but I usually try to avoid I/O in constructors. From past experience, effectful constructors can make testing harder, and, even when called out (like you've done here!), can be surprising for callers.
* The network request can take an unknown amount of time, blocking the thread that creates the `Client` until it completes. Consumers of the client will need to remember to construct it lazily, and off the main thread. The complexity with mutexes / optionals / `OnceCell`s will still be there, but pushed to the consumers.
* If the request fails—maybe the server had a hiccup, or the user is on a spotty network—we'll fail to create the client.
* Every consumer of `Client` will need to wait for the network call, even if they're not interested in attachments.

It's more work for us to look up the base URL lazily, the first time `get_attachment` is called, but it would mean that:

* Consumers can call `Client::new` on the main / any thread, and not worry about it blocking.
* `Client::new` errors are much more deterministic: the only reason it'll fail is if the caller gives it a bad server URL, not a potentially transient network failure.
* Consumers that never call `get_attachment` won't make the extra network call at all!

WDYT? Would you be open to discussing more, or changing the behavior to be lazy?

Reply:
You're totally right, I (ironically) was just being lazy. I agree that the the lazy implementation is the more correct way to handle this. I'll update accordingly. I do really appreciate the detailed write-up!